### PR TITLE
refactor(emqx_mgmt): Decorate RPCs

### PR DESCRIPTION
**Pull request by k32**

### CODE REVIEW

File: apps/emqx/src/proto/emqx_proto_v1.erl [hunk 1]
```diff
@@ -30,3 +37,22 @@ introduced_in() ->
 -spec is_running(node()) -> boolean() | {badrpc, term()}.
 is_running(Node) ->
     rpc:call(Node, emqx, is_running, []).
+
+-spec get_stats(node()) -> emqx_stats:stats() | {badrpc, _}.
```
Comment (thalesmg) [inline apps/emqx/src/proto/emqx_proto_v1.erl:41]: Q: would `emqx_rpc:badrpc()` be more precise for the error type?

File: apps/emqx_management/src/emqx_mgmt.erl [hunk 2]
```diff
@@ -593,6 +578,11 @@ rpc_call(Node, Fun, Args) ->
         Res -> Res
     end.
```
Comment (thalesmg) [inline apps/emqx_management/src/emqx_mgmt.erl:579]: Suggestion: maybe we could remove this `rpc_call/3` function.

File: apps/emqx_management/src/proto/emqx_management_proto_v1.erl [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+broker_info(Node) ->
+    rpc:call(Node, emqx_mgmt, broker_info, []).
+
+-spec list_subscriptions(node()) -> [map()] | {badrpc, _}.
+list_subscriptions(Node) ->
+    rpc:call(Node, emqx_mgmt, do_list_subscriptions, []).
+
+-spec kickout_client(node(), emqx_types:clientid()) -> ok | {badrpc, _}.
+kickout_client(Node, ClientId) ->
+    rpc:call(Node, emqx_cm, kick_session, [ClientId]).
```
Comment (thalesmg) [inline apps/emqx_management/src/proto/emqx_management_proto_v1.erl:51]: Q: I noticed that a few calls that were made in `mgmt` (like `get_metrics` for example) went to `emqx_proto`, but this one is here in `emqx_management_proto`.  If we are separating the API by the app that owns the module being called, shouldn't this one be in `emqx_proto`?

File: apps/emqx_management/src/emqx_mgmt.erl [hunk 4]
```diff
@@ -593,6 +578,11 @@ rpc_call(Node, Fun, Args) ->
...
 
+wrap_rpc({badrpc, Reason}) ->
```
Comment (thalesmg) [inline apps/emqx_management/src/emqx_mgmt.erl:581]: Could be `{badtcp, _}` as well?

File: apps/emqx_management/src/emqx_mgmt.erl [hunk 5]
```diff
@@ -455,12 +440,12 @@ do_unsubscribe(ClientId, Topic) ->
...
+    [Conf#{node => node(), id => Id} || {Id, Conf} <- emqx_listeners:list()].
+
 list_listeners() ->
     lists:append([list_listeners(Node) || Node <- mria_mnesia:running_nodes()]).
 
-list_listeners(Node) when Node =:= node() ->
-    [Conf#{node => Node, id => Id} || {Id, Conf} <- emqx_listeners:list()];
-
 list_listeners(Node) ->
     rpc_call(Node, list_listeners, [Node]).
```
Comment (thalesmg) [inline apps/emqx_management/src/emqx_mgmt.erl:450]: This one will get into an infinite loop?

File: apps/emqx/src/proto/emqx_proto_v1.erl (hunk 1 above)
Comment (k32) [reply apps/emqx/src/proto/emqx_proto_v1.erl:41]: Not in this case, `emqx_rpc:badrpc()` is the error type for `gen_rpc`, for `rpc`: it's just https://www.erlang.org/doc/man/rpc.html#call-4

File: apps/emqx_management/src/emqx_mgmt.erl (hunk 2 above)
Comment (k32) [reply apps/emqx_management/src/emqx_mgmt.erl:579]: Yes, that's the intention, but the rest of the work here requires more thorough refactoring.

File: apps/emqx_management/src/proto/emqx_management_proto_v1.erl (hunk 3 above)
Comment (k32) [reply apps/emqx_management/src/proto/emqx_management_proto_v1.erl]: Good call, fixed.

File: apps/emqx/src/emqx_cm.erl [hunk 6]
```diff
@@ -502,6 +512,18 @@ lookup_channels(global, ClientId) ->
 lookup_channels(local, ClientId) ->
     [ChanPid || {_, ChanPid} <- ets:lookup(?CHAN_TAB, ClientId)].
 
+-spec lookup_client({clientid, emqx_types:clientid()} | {username, emqx_types:username()}) ->
+          [channel_info()].
+lookup_client({username, Username}) ->
+    MatchSpec = [{ {'_', #{clientinfo => #{username => '$1'}}, '_'}
```
Comment (savonarola) [inline apps/emqx/src/emqx_cm.erl:518]: nit: is there a specific reason for not using `fun2ms`?

**Review by qzhuyan — APPROVED** [review]

**Review by savonarola — APPROVED** [review]

File: apps/emqx/src/emqx_cm.erl (hunk 6 above)
Comment (k32) [reply apps/emqx/src/emqx_cm.erl:518]: This is copy-paste of the old code, I try to minimize business logic refactoring in these PRs to save time.

**Pull request merged**